pkg/vault/contracts/Vault.sol

```diff
@@ -1148,44 +1146,55 @@ contract Vault is IVaultMain, VaultCommon, Proxy {
             uint256 bufferUnderlyingSurplus = bufferBalances.getBufferUnderlyingSurplus(wrappedToken);
             uint256 bufferWrappedSurplus;
 
-            if (bufferUnderlyingSurplus > 0) {
-                bufferWrappedSurplus = wrappedToken.convertToShares(bufferUnderlyingSurplus);
-            }
-
-            uint256 calculatedUnderlyingDelta;
-            uint256 calculatedWrappedDelta;
+            uint256 vaultUnderlyingDelta;
+            uint256 vaultWrappedDelta;
 
             if (kind == SwapKind.EXACT_IN) {
                 // The amount of underlying tokens to deposit is the necessary amount to fulfill the trade
                 // (amountInUnderlying), plus the amount needed to leave the buffer rebalanced 50/50 at the end
                 // (bufferUnderlyingSurplus).
-                calculatedUnderlyingDelta = amountInUnderlying + bufferUnderlyingSurplus;
-
-                underlyingToken.forceApprove(address(wrappedToken), calculatedUnderlyingDelta);
-                // EXACT_IN requires the exact amount of underlying tokens to be deposited, so deposit is called.
-                wrappedToken.deposit(calculatedUnderlyingDelta, address(this));
+                vaultUnderlyingDelta = amountInUnderlying + bufferUnderlyingSurplus;
+                if (isQueryContext && underlyingToken.balanceOf(address(this)) < vaultUnderlyingDelta) {
+                    vaultWrappedDelta = wrappedToken.previewDeposit(vaultUnderlyingDelta);
+                } else {
+                    underlyingToken.forceApprove(address(wrappedToken), vaultUnderlyingDelta);
+                    // EXACT_IN requires the exact amount of underlying tokens to be deposited, so deposit is called.
+                    vaultWrappedDelta = wrappedToken.deposit(vaultUnderlyingDelta, address(this));
+                }
             } else {
-                // Note that `bufferWrappedSurplus` will be zero if there is no bufferUnderlyingSurplus.
-                calculatedWrappedDelta = amountOutWrapped + bufferWrappedSurplus;
+                if (bufferUnderlyingSurplus > 0) {
+                    bufferWrappedSurplus = wrappedToken.convertToShares(bufferUnderlyingSurplus);
+                }
+                vaultWrappedDelta = amountOutWrapped + bufferWrappedSurplus;
+
+                if (
+                    isQueryContext &&
+                    underlyingToken.balanceOf(address(this)) < amountInUnderlying + bufferUnderlyingSurplus
+                ) {
+                    vaultUnderlyingDelta = wrappedToken.previewMint(vaultWrappedDelta);
+                } else {
+                    // Add convert error because mint can consume a different amount of tokens than we anticipated.
+                    underlyingToken.forceApprove(
+                        address(wrappedToken),
+                        _addConvertError(amountInUnderlying + bufferUnderlyingSurplus)
+                    );
```

Comment:
Not specific to this PR, but maybe this was related to the old wrappers we were using?

According to https://eips.ethereum.org/EIPS/eip-4626, `mint` must mint exact shares. Now that we're reviewing these details, we might save a few bytes here :)

Reply:
Hmm, mint is a confusing function:
* We ask to mint a certain amount of shares
* The wrapper gets underlying, and return the exact amount of shares.

Here we approve the underlying tokens that mint will consume, which is not an exact number. We use amountInUnderlying and bufferUnderlyingSurplus, but they were calculated by convertToAssets, so they have errors.

Reply:
got it, thanks for the explanation! Can we add that detail to the comments please?

Reply:
Done!